torch/_inductor/ir.py

```diff
@@ -3318,6 +3318,249 @@ def get_template_tiling(self):
             sympy.Integer(1),
         )
 
+def get_num_ranks(arg):
+    # hack, assume we're using arg (pg_name) as number of ranks for now.
+    return int(arg)
+
+
+
+# HACK
+# Takes 1 tensor in input, produces 1 tensor in output
+class ReduceScatterTensor(ExternKernelOut):
+    def __init__(
+        self,
+        layout,
+        inputs,
+        constant_args=(),
+    ):
+        super().__init__(layout, inputs, constant_args)
+
+    @classmethod
+    def create(
+        cls,
+        x: "TensorBox",
+        arg: str,
+    ):
+        size = x.get_size()
+        x = cls.realize_input(cls.require_contiguous(x))
+        num_ranks = get_num_ranks(arg)
+        out_layout = FixedLayout(
+            x.get_device(),
+            x.get_dtype(),
+            [size[0] // num_ranks] + size[1:],
+            x.get_stride(),
+        )
+        return ReduceScatterTensor(
+            layout=out_layout,
+            inputs=[x],
+        )
+
+    def codegen(self, wrapper):
+        wrapper.header.writeline(
+            f"import torch.distributed as dist"
+        )
+        wrapper.writeline(
+            f"{self.get_name()}_work = dist.reduce_scatter_tensor({self.get_name()}, {self.codegen_args()[0]}, async_op=True)"
+        )
+        wrapper.writeline(
+            f"{self.get_name()}_work.wait()"
+        )
+        wrapper.writeline(
+            f"{self.get_name()} = {self.codegen_args()[0]}"
+        )
+
+
+# HACK
+# Takes 1 tensor in input, produces world_size tensors in output
+class AllGather(ExternKernelOut):
+    def __init__(
+        self,
+        layout,
+        inputs,
+        constant_args=(),
+    ):
+        super().__init__(layout, inputs, constant_args)
+
+    @classmethod
+    def create(
+        cls,
+        x: "TensorBox",
+        arg: str,
+    ):
+        x = cls.realize_input(cls.require_contiguous(x))
+        out_layout = FixedLayout(
+            x.get_device(),
+            x.get_dtype(),
+            x.size(),
+            x.stride(),
+        )
+        return AllGather(
+            layout=MultiOutputLayout(x.get_device()),
```

Comment:
@wconstab i'm completely lost here on how to do it.

AllGather returns multiple tensors so i need to allocate multiiple buffers and hten pass then to Allgather as a list of ouptuts.

ExternKernelOut only creates 1 buffer for me , so not sure how to do it .

Reply:
I think one way to do it is that since the current traceable op would take the 'output' tensorlist as inputs, the traced graph would already include a call to allocate those empty tensors before the allgather.  we could make allgather a 'mutation' op just like allreduce, that mutated even more buffers.  but, i'm nt sure if this is really what we want to do as we are implementing a new functional IR.  I will think about this more and try to figure out the right way